cmd/frontend/graphqlbackend/repositories_test.go

```diff
@@ -689,114 +708,295 @@ func TestRepositories_Integration(t *testing.T) {
 	ctx = actor.WithActor(ctx, actor.FromUser(admin.ID))
 
 	tests := []repositoriesQueryTest{
-		// no args
+		// first
 		{
-			wantRepos:      []string{"repo1", "repo2", "repo3", "repo4", "repo5", "repo6", "repo7", "repo8"},
-			wantTotalCount: 8,
+			args:             "first: 2",
+			wantRepos:        []string{"repo1", "repo2"},
+			wantTotalCount:   8,
+			wantNextPage:     true,
+			wantPreviousPage: false,
+			wantStartCursor:  buildCursor(repos[0].repo),
+			wantEndCursor:    buildCursor(repos[1].repo),
 		},
-		// first
+		// second page with first, after args
+		{
+			args:             fmt.Sprintf(`first: 2, after: "%s"`, *buildCursor(repos[0].repo)),
+			wantRepos:        []string{"repo2", "repo3"},
+			wantTotalCount:   8,
+			wantNextPage:     true,
+			wantPreviousPage: true,
+			wantStartCursor:  buildCursor(repos[1].repo),
+			wantEndCursor:    buildCursor(repos[2].repo),
+		},
+		// last page with first, after args
+		{
+			args:             fmt.Sprintf(`first: 2, after: "%s"`, *buildCursor(repos[5].repo)),
+			wantRepos:        []string{"repo7", "repo8"},
+			wantTotalCount:   8,
+			wantNextPage:     false,
+			wantPreviousPage: true,
+			wantStartCursor:  buildCursor(repos[6].repo),
+			wantEndCursor:    buildCursor(repos[7].repo),
+		},
+		// last
+		{
+			args:             "last: 2",
+			wantRepos:        []string{"repo7", "repo8"},
+			wantTotalCount:   8,
+			wantNextPage:     false,
+			wantPreviousPage: true,
+			wantStartCursor:  buildCursor(repos[6].repo),
+			wantEndCursor:    buildCursor(repos[7].repo),
+		},
+		// second last page with last, before args
+		{
+			args:             fmt.Sprintf(`last: 2, before: "%s"`, *buildCursor(repos[6].repo)),
+			wantRepos:        []string{"repo5", "repo6"},
+			wantTotalCount:   8,
+			wantNextPage:     true,
+			wantPreviousPage: true,
+			wantStartCursor:  buildCursor(repos[4].repo),
+			wantEndCursor:    buildCursor(repos[5].repo),
+		},
+		// back to first page with last, before args
+		{
+			args:             fmt.Sprintf(`last: 2, before: "%s"`, *buildCursor(repos[2].repo)),
+			wantRepos:        []string{"repo1", "repo2"},
+			wantTotalCount:   8,
+			wantNextPage:     true,
+			wantPreviousPage: false,
+			wantStartCursor:  buildCursor(repos[0].repo),
+			wantEndCursor:    buildCursor(repos[1].repo),
+		},
+		// descending first
+		{
+			args:             "first: 2, descending: true",
+			wantRepos:        []string{"repo8", "repo7"},
+			wantTotalCount:   8,
+			wantNextPage:     true,
+			wantPreviousPage: false,
+			wantStartCursor:  buildCursor(repos[7].repo),
+			wantEndCursor:    buildCursor(repos[6].repo),
+		},
+		// descending second page with first, after args
+		{
+			args:             fmt.Sprintf(`first: 2, descending: true, after: "%s"`, *buildCursor(repos[6].repo)),
```

Comment:
Nit: you can use `%q` instead of `"%s"` because it will *q*uote the string :)

```suggestion
			args:             fmt.Sprintf(`first: 2, descending: true, after: "%s"`, *buildCursor(repos[6].repo)),
```